File: src/lidar_camera_fusion.cpp
#include "ros2_edge_perception/lidar_camera_fusion.hpp"
#include <numeric>
#include <cmath>
#include <limits>
#include <string>

namespace ros2_edge_perception {

LidarCameraFusionEngine::LidarCameraFusionEngine(const Intrinsics& intrinsics,
                                                 const Extrinsics& extrinsics)
    : intrinsics_(intrinsics), extrinsics_(extrinsics) {}

void LidarCameraFusionEngine::set_intrinsics(const Intrinsics& intrinsics) {
    intrinsics_ = intrinsics;
}

void LidarCameraFusionEngine::set_extrinsics(const Extrinsics& extrinsics) {
    extrinsics_ = extrinsics;
}

bool LidarCameraFusionEngine::project_point(const LidarPoint& pt_lidar,
                                            float& u, float& v, float& depth_cam) const {
    // 1. Transform LiDAR 3D point into Camera Optical Frame: p_cam = R * p_lidar + t
    Eigen::Vector3f p_l(pt_lidar.x, pt_lidar.y, pt_lidar.z);
    Eigen::Vector3f p_c = extrinsics_.R * p_l + extrinsics_.t;

    // Reject points behind or too close to camera plane
    if (p_c.z() <= 0.1f) {
        return false;
    }

    depth_cam = p_c.z();

    // 2. Pin-Hole Perspective Projection
    u = (intrinsics_.fx * p_c.x()) / depth_cam + intrinsics_.cx;
    v = (intrinsics_.fy * p_c.y()) / depth_cam + intrinsics_.cy;

    return true;
}

std::vector<FusedDetection3D> LidarCameraFusionEngine::fuse(
    const std::vector<BBox2D>& detections_2d,
    const std::vector<LidarPoint>& lidar_points,
    float min_depth,
    float max_depth) const {

    std::vector<FusedDetection3D> fused_results;
    if (detections_2d.empty()) {
        return fused_results;
    }

    // Structure to hold projected point in camera space
    struct ProjectedPoint {
        float u, v;
        Eigen::Vector3f p_c;
    };

    std::vector<ProjectedPoint> valid_proj_points;
    valid_proj_points.reserve(lidar_points.size());

    for (const auto& pt : lidar_points) {
        float u, v, d;
        if (project_point(pt, u, v, d)) {
            if (d >= min_depth && d <= max_depth) {
                Eigen::Vector3f p_l(pt.x, pt.y, pt.z);
                Eigen::Vector3f p_c = extrinsics_.R * p_l + extrinsics_.t;
                valid_proj_points.push_back({u, v, p_c});
            }
        }
    }

    for (const auto& det : detections_2d) {
        FusedDetection3D fused;
        fused.class_name = det.class_name;
        fused.class_id = det.class_id;
        fused.score = det.score;
        fused.bbox_2d_center_x = (det.x1 + det.x2) / 2.0f;
        fused.bbox_2d_center_y = (det.y1 + det.y2) / 2.0f;
        fused.bbox_2d_width = det.x2 - det.x1;
        fused.bbox_2d_height = det.y2 - det.y1;

        // Collect all LiDAR points falling inside this 2D bounding box
        std::vector<Eigen::Vector3f> box_points;
        std::vector<float> depths;

        for (const auto& proj : valid_proj_points) {
            if (proj.u >= det.x1 && proj.u <= det.x2 &&
                proj.v >= det.y1 && proj.v <= det.y2) {
                box_points.push_back(proj.p_c);
                depths.push_back(proj.p_c.z());
            }
        }

        // Tier-1 Statistical Clustering: Need at least 3 points for confidence
        if (depths.size() >= 3) {
            std::sort(depths.begin(), depths.end());
            float median_depth = depths[depths.size() / 2];

            // Filter points within 15% distance window around median depth (eliminating background/foreground pass-through)
            float depth_window = std::max(0.4f, median_depth * 0.15f);
            std::vector<Eigen::Vector3f> clustered_points;

            for (const auto& p : box_points) {
                if (std::abs(p.z() - median_depth) <= depth_window) {
                    clustered_points.push_back(p);
                }
            }

            if (!clustered_points.empty()) {
                float min_x = clustered_points[0].x(), max_x = clustered_points[0].x();
                float min_y = clustered_points[0].y(), max_y = clustered_points[0].y();
                float min_z = clustered_points[0].z(), max_z = clustered_points[0].z();

                for (const auto& p : clustered_points) {
                    min_x = std::min(min_x, p.x());
                    max_x = std::max(max_x, p.x());
                    min_y = std::min(min_y, p.y());
                    max_y = std::max(max_y, p.y());
                    min_z = std::min(min_z, p.z());
                    max_z = std::max(max_z, p.z());
                }

                fused.x = (min_x + max_x) / 2.0f;
                fused.y = (min_y + max_y) / 2.0f;
                fused.z = (min_z + max_z) / 2.0f;

                fused.size_x = std::max(0.3f, max_x - min_x);
                fused.size_y = std::max(0.3f, max_y - min_y);
                fused.size_z = std::max(0.3f, max_z - min_z);

                // Estimate Heading (Yaw) using principal component of X-Z points
                if (clustered_points.size() >= 5) {
                    float mean_x = fused.x;
                    float mean_z = fused.z;
                    float cov_xx = 0.0f, cov_xz = 0.0f, cov_zz = 0.0f;
                    for (const auto& p : clustered_points) {
                        float dx = p.x() - mean_x;
                        float dz = p.z() - mean_z;
                        cov_xx += dx * dx;
                        cov_xz += dx * dz;
                        cov_zz += dz * dz;
                    }
                    // Heading angle in ground plane (X-Z)
                    fused.yaw = 0.5f * std::atan2(2.0f * cov_xz, cov_xx - cov_zz);
                    fused.qx = 0.0f;
                    fused.qy = std::sin(fused.yaw / 2.0f); // Rotation around camera Y (vertical)
                    fused.qz = 0.0f;
                    fused.qw = std::cos(fused.yaw / 2.0f);
                }

                fused.lidar_point_count = clustered_points.size();
                fused.fused_with_lidar = true;
                fused.is_valid_3d = true;
                fused.geometry_status = "LIDAR_FUSED";
                fused_results.push_back(fused);
                continue;
            }
        }

        // Fallback if no LiDAR points available inside this box:
        // Production safety: mark explicitly as unmeasured rather than fabricating arbitrary 3.0m geometry.
        fused.x = std::numeric_limits<float>::quiet_NaN();
        fused.y = std::numeric_limits<float>::quiet_NaN();
        fused.z = std::numeric_limits<float>::quiet_NaN();
        fused.size_x = 0.0f;
        fused.size_y = 0.0f;
        fused.size_z = 0.0f;
        fused.fused_with_lidar = false;
        fused.is_valid_3d = false;
        fused.geometry_status = "INVALID_UNMEASURED_DEPTH";
        fused_results.push_back(fused);
    }

    return fused_results;
}

} // namespace ros2_edge_perception

File: ros2_edge_perception/lidar_camera_fusion.py
"""
LiDAR-Camera Multi-Sensor Fusion Engine for ROS 2.

Features:
1. 3D Frustum Projection: Projects LiDAR PointCloud (X_L, Y_L, Z_L) to Camera Pixel Plane (u, v).
2. 2D Bounding Box Gating: Associates LiDAR points falling inside YOLO 2D detections.
3. Statistical Depth Clustering: Median filtering & outlier rejection eliminating infrared sunlight washout.
4. Heading (Yaw) & Quaternion Estimation: Computes 3D oriented bounding boxes.
"""

import math
from typing import Dict, List, Optional, Tuple
import numpy as np


class LidarCameraFusionEngine:
    """Fuses 2D visual detections with 3D LiDAR point clouds."""

    def __init__(
        self,
        fx: float = 554.25,
        fy: float = 554.25,
        cx: float = 320.0,
        cy: float = 240.0,
        extrinsic_rot: Optional[np.ndarray] = None,
        extrinsic_trans: Optional[np.ndarray] = None,
    ):
        self.fx = fx
        self.fy = fy
        self.cx = cx
        self.cy = cy

        # Default Extrinsics: Identity rotation, zero translation
        self.R = extrinsic_rot if extrinsic_rot is not None else np.eye(3, dtype=np.float32)
        self.t = extrinsic_trans if extrinsic_trans is not None else np.zeros(3, dtype=np.float32)

    def set_intrinsics(self, fx: float, fy: float, cx: float, cy: float):
        self.fx = fx
        self.fy = fy
        self.cx = cx
        self.cy = cy

    def set_extrinsics(self, R: np.ndarray, t: np.ndarray):
        self.R = R.astype(np.float32)
        self.t = t.astype(np.float32)

    def project_points(self, lidar_points: np.ndarray) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
        """
        Project N x 3 LiDAR points into 2D camera pixels (u, v) and metric depth z_c.
        Returns:
            u: array of horizontal pixel coordinates
            v: array of vertical pixel coordinates
            z_c: array of metric depths in camera optical frame
        """
        if len(lidar_points) == 0:
            return np.array([]), np.array([]), np.array([])

        # 1. Transform points: P_c = R * P_l + t
        pts_c = (self.R @ lidar_points[:, :3].T).T + self.t

        # Filter points in front of camera plane (Z > 0.1m)
        valid_mask = pts_c[:, 2] > 0.1
        if not np.any(valid_mask):
            return np.array([]), np.array([]), np.array([])

        pts_c_valid = pts_c[valid_mask]
        z_c = pts_c_valid[:, 2]

        # 2. Pin-Hole Perspective Projection
        u = (self.fx * pts_c_valid[:, 0]) / z_c + self.cx
        v = (self.fy * pts_c_valid[:, 1]) / z_c + self.cy

        return u, v, pts_c_valid

    def fuse(
        self,
        detections_2d: List[Dict],
        lidar_points: np.ndarray,
        min_depth: float = 0.5,
        max_depth: float = 80.0,
    ) -> List[Dict]:
        """
        Fuses 2D bounding boxes with 3D LiDAR point clouds.
        Returns list of 3D fused obstacle dictionaries with metric bounds and orientation quaternions.
        """
        fused_results = []
        if not detections_2d:
            return fused_results

        if len(lidar_points) > 0:
            u, v, pts_c = self.project_points(lidar_points)
        else:
            u, v, pts_c = np.array([]), np.array([]), np.array([])

        for det in detections_2d:
            fused = {
                "class_name": det["class_name"],
                "class_id": det["class_id"],
                "score": det["score"],
                "x": 0.0,
                "y": 0.0,
                "z": 3.0,
                "size_x": 0.5,
                "size_y": 0.5,
                "size_z": 0.5,
                "yaw": 0.0,
                "qx": 0.0,
                "qy": 0.0,
                "qz": 0.0,
                "qw": 1.0,
                "lidar_point_count": 0,
                "fused_with_lidar": False,
            }

            x1, y1, x2, y2 = det["x1"], det["y1"], det["x2"], det["y2"]

            # Association with LiDAR points
            if len(u) > 0:
                in_box_mask = (u >= x1) & (u <= x2) & (v >= y1) & (v <= y2)
                in_box_pts = pts_c[in_box_mask]

                # Distance range gating
                valid_depth_mask = (in_box_pts[:, 2] >= min_depth) & (in_box_pts[:, 2] <= max_depth)
                gated_pts = in_box_pts[valid_depth_mask]

                if len(gated_pts) >= 3:
                    depths = gated_pts[:, 2]
                    median_depth = np.median(depths)
                    depth_window = max(0.4, median_depth * 0.15)

                    # Cluster filtering around median depth
                    cluster_mask = np.abs(depths - median_depth) <= depth_window
                    clustered = gated_pts[cluster_mask]

                    if len(clustered) >= 2:
                        min_bounds = np.min(clustered, axis=0)
                        max_bounds = np.max(clustered, axis=0)

                        fused["x"] = float((min_bounds[0] + max_bounds[0]) / 2.0)
                        fused["y"] = float((min_bounds[1] + max_bounds[1]) / 2.0)
                        fused["z"] = float((min_bounds[2] + max_bounds[2]) / 2.0)

                        fused["size_x"] = float(max(0.3, max_bounds[0] - min_bounds[0]))
                        fused["size_y"] = float(max(0.3, max_bounds[1] - min_bounds[1]))
                        fused["size_z"] = float(max(0.3, max_bounds[2] - min_bounds[2]))

                        # Heading (Yaw) and Quaternion calculation
                        if len(clustered) >= 5:
                            dx = clustered[:, 0] - fused["x"]
                            dz = clustered[:, 2] - fused["z"]
                            cov_xx = np.mean(dx * dx)
                            cov_xz = np.mean(dx * dz)
                            cov_zz = np.mean(dz * dz)
                            yaw = 0.5 * math.atan2(2.0 * cov_xz, cov_xx - cov_zz)
                            fused["yaw"] = float(yaw)
                            fused["qy"] = float(math.sin(yaw / 2.0))
                            fused["qw"] = float(math.cos(yaw / 2.0))

                        fused["lidar_point_count"] = len(clustered)
                        fused["fused_with_lidar"] = True
                        fused_results.append(fused)
                        continue

            # Fallback to pin-hole box center
            cx_box = (x1 + x2) / 2.0
            cy_box = (y1 + y2) / 2.0
            fallback_depth = det.get("depth", 3.0)
            fused["x"] = float((cx_box - self.cx) * fallback_depth / self.fx)
            fused["y"] = float((cy_box - self.cy) * fallback_depth / self.fy)
            fused["z"] = float(fallback_depth)
            fused["size_x"] = float(max(0.4, (x2 - x1) * fallback_depth / self.fx))
            fused["size_y"] = float(max(0.4, (y2 - y1) * fallback_depth / self.fy))
            fused["size_z"] = 0.5
            fused_results.append(fused)

        return fused_results

